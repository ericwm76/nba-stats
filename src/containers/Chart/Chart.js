import React, { Component } from 'react';
import { Line } from 'react-chartjs-2';
import PropTypes from 'prop-types';

class Chart extends Component {
  constructor() {
    super();
    this.state = {
      labels: [],
      datasets: []
    }
  }

  createLabels(stats) {
    let n = 0;
    let labels = [];

    stats.forEach(stat => {
      let id = Object.keys(stat);
      if (stat[id].length > n) {
        n = stat[id].length
      }
    })

    for (var i = 0; i < n; i++) {
      labels.push("Game " + (i + 1))
    }

    this.setState({ labels: labels })
  }

  createDatasets(players, stats) {
    let datasets = [];

    stats.forEach(stat => {
      let id = parseInt(Object.keys(stat));
      let player = players.find(player => player.id === id);
      let playerName = player.first_name + ' ' + player.last_name;
      let points = stat[id].map(game => game.pts);

      datasets.push({
        label: playerName,
        fill: false,
        lineTension: 0.1,
        backgroundColor: 'rgba(75,192,192,0.4)',
        borderColor: 'rgba(75,192,192,1)',
        borderCapStyle: 'butt',
        borderDash: [],
        borderDashOffset: 0.0,
        borderJoinStyle: 'miter',
        pointBorderColor: 'rgba(75,192,192,1)',
        pointBackgroundColor: '#fff',
        pointBorderWidth: 1,
        pointHoverRadius: 5,
        pointHoverBackgroundColor: 'rgba(75,192,192,1)',
        pointHoverBorderColor: 'rgba(220,220,220,1)',
        pointHoverBorderWidth: 2,
        pointRadius: 1,
        pointHitRadius: 10,
        data: points
      })
    })

    this.setState({ datasets: datasets });
  }

  componentDidMount() {
    const { stats, players } = this.props
    this.createLabels(stats);
    this.createDatasets(players, stats);
  }

  render() {
    const { labels, datasets } = this.state

    let data = {
      labels: labels.length > 10 ? labels.slice(-10) : labels,
      datasets: datasets
    }

    return (
      <div>
        <h2>Points Per Game - Last 10 Games</h2>
        <Line data={data} />
      </div>
    );
  }
};

export default Chart;

Chart.propTypes = {
  players: PropTypes.array.isRequired,
  stats: PropTypes.array.isRequired
}