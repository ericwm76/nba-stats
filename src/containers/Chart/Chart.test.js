import React from 'react';
import { shallow } from 'enzyme';
import { Chart } from './Chart';

describe('Chart', () => {
  let wrapper;
  const mockPlayers = [{
    first_name: "Nikola",
    last_name: "Jokic",
    height_feet: 7,
    height_inches: 0,
    weight_pounds: 280,
    position: "C",
    id: 246,
    team: {
      full_name: "Denver Nuggets",
    }
  },
  {
    first_name: "Lebron",
    last_name: "James",
    height_feet: 6,
    height_inches: 8,
    weight_pounds: 250,
    position: "F",
    id: 237,
    team: {
      full_name: "Los Angeles Lakers",
    }
  }];

  const mockStats = [
    { 237: [
      { 0: 
        {
        "id":29,
        "ast":2,
        "blk":2,
        "dreb":8,
        "fg3_pct":0.25,
        "fg3a":4,
        "fg3m":1,
        "fg_pct":0.429,
        "fga":21,
        "fgm":9,
        "ft_pct":0.8,
        "fta":5,
        "ftm":4,
        "game":{
          "id":1,
          "date":"2018-10-16T00:00:00.000Z",
          "home_team_id":2,
          "home_team_score":105,
          "season":2018,
          "visitor_team_id":23,
          "visitor_team_score":87
        },
        "min":"36:49",
        "oreb":2,
        "pf":3,
        "player":{
          "id":246,
          "first_name":"Lebron",
          "last_name":"James",
          "position":"F",
          "team_id":23
        },
        "pts":23,
        "reb":10,
        "stl":1,
        "team":{
          "id":23,
          "abbreviation":"LAL",
          "city":"Los Angeles",
          "conference":"West",
          "division":"Pacific",
          "full_name":"Los Angeles Lakers",
          "name":"Lakers"
        },
        "turnover":5
          }
        }
      ]
    },
    { 246: [
      { 0: 
        {
        "id":29,
        "ast":2,
        "blk":2,
        "dreb":8,
        "fg3_pct":0.25,
        "fg3a":4,
        "fg3m":1,
        "fg_pct":0.429,
        "fga":21,
        "fgm":9,
        "ft_pct":0.8,
        "fta":5,
        "ftm":4,
        "game":{
          "id":1,
          "date":"2018-10-16T00:00:00.000Z",
          "home_team_id":2,
          "home_team_score":105,
          "season":2018,
          "visitor_team_id":23,
          "visitor_team_score":87
        },
        "min":"36:49",
        "oreb":2,
        "pf":3,
        "player":{
          "id":246,
          "first_name":"Nikola",
          "last_name":"Jokic",
          "position":"C",
          "team_id":23
        },
        "pts":23,
        "reb":10,
        "stl":1,
        "team":{
          "id":23,
          "abbreviation":"DEN",
          "city":"Denver",
          "conference":"West",
          "division":"Mountain",
          "full_name":"Denver Nuggets",
          "name":"Nuggets"
        },
        "turnover":5
          }
        }
      ]
    }
  ]
  
  beforeEach(() => {
    wrapper = shallow(<Chart stats={mockStats} players={mockPlayers}/>);
    wrapper.instance().createLabels = jest.fn();    
  })

  it('should call createLabels and createDatasets on mount', () => {
    wrapper.instance().componentDidMount();

    expect(wrapper.instance().createLabels).toHaveBeenCalled();
  })

  it('should update labels in state when createLabels is called', () => {
  })
})