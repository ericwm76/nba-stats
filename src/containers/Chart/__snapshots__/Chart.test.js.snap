// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart should match the snapshot 1`] = `
<div>
  <h2>
    Points Per Game - Last 10 Games
  </h2>
  <Line
    data={
      Object {
        "datasets": Array [
          Object {
            "backgroundColor": "rgba(75,192,192,0.4)",
            "borderCapStyle": "butt",
            "borderColor": "rgba(75,192,192,1)",
            "borderDash": Array [],
            "borderDashOffset": 0,
            "borderJoinStyle": "miter",
            "data": Array [
              undefined,
            ],
            "fill": false,
            "label": "Lebron James",
            "lineTension": 0.1,
            "pointBackgroundColor": "#fff",
            "pointBorderColor": "rgba(75,192,192,1)",
            "pointBorderWidth": 1,
            "pointHitRadius": 10,
            "pointHoverBackgroundColor": "rgba(75,192,192,1)",
            "pointHoverBorderColor": "rgba(220,220,220,1)",
            "pointHoverBorderWidth": 2,
            "pointHoverRadius": 5,
            "pointRadius": 1,
          },
          Object {
            "backgroundColor": "rgba(75,192,192,0.4)",
            "borderCapStyle": "butt",
            "borderColor": "rgba(75,192,192,1)",
            "borderDash": Array [],
            "borderDashOffset": 0,
            "borderJoinStyle": "miter",
            "data": Array [
              undefined,
            ],
            "fill": false,
            "label": "Nikola Jokic",
            "lineTension": 0.1,
            "pointBackgroundColor": "#fff",
            "pointBorderColor": "rgba(75,192,192,1)",
            "pointBorderWidth": 1,
            "pointHitRadius": 10,
            "pointHoverBackgroundColor": "rgba(75,192,192,1)",
            "pointHoverBorderColor": "rgba(220,220,220,1)",
            "pointHoverBorderWidth": 2,
            "pointHoverRadius": 5,
            "pointRadius": 1,
          },
        ],
        "labels": Array [
          "Game 1",
        ],
      }
    }
  />
</div>
`;
